Correct the handling of keyword arguments in Derivative.__new__

normalhuman opened the pull request:
#### References to other Issues or PRs

Fixes #12996

#### Brief description of what is fixed or changed

Currently, the `__new__` method of Derivative accepts keyword arguments, handles two of them (evaluate and simplify) and tries to pass the rest to any new derivatives created in the process, which raises an error because there are no keyword arguments for `Expr.__new__`. This is corrected now: it no longer passes the rest to `Expr.__new__`.

#### Other comments

We cannot yet change the method signature to `__new__(cls, expr, *variables, evaluate=False, simplify=True`) because Python 2.7 does not support keyword-only arguments appearing after `*args`.

The bug manifested itself in #12996 because a derivative was created in the process of integration, and then received `manual=False` parameter (intended for Integral) which it tried to pass to `Expr.__new__`.

Thanks to @asmeurer for proposing this solution in Gitter chat.

normalhuman committed:
Correct handling of keyword arguments in Derivative.doit

Fixes #12996

Currently, the doit method of Derivative accepts keyword arguments, handles
two of them (evaluate and simplify) and tries to pass the rest to any new
derivatives created in the process, which raises an error because there are
no keyword arguments for `Expr.__new__`. This is corrected now.

The bug manifested itself in #12996 because a derivative was created in the
process of integration, and then received manual=False parameter (intended for
Integral) which it tried to pass to `Expr.__new__`.

normalhuman committed:
The .pop method no longer serves a purpose here, replaced by .get

asmeurer commented on sympy/integrals/tests/test_integrals.py:
```diff
@@ -1166,6 +1166,10 @@ def test_issue_8901():
     assert integrate(tanh(1.0*x)) == 1.0*x - 1.0*log(tanh(1.0*x) + 1)
     assert integrate(tanh(x)) == x - log(tanh(x) + 1)
 
+@slow
```
How slow is this test?

normalhuman replied:
timeit measured it at 23.4 seconds on my EC2 instance (nano). I found it to be a lot of time to wait for an integral to come back unevaluated.

ylemkimon replied:
I think just passing any keyword argument to Derivative is enough for the test.

normalhuman replied:
I agree; since this was the actual cause of the error, I replaced the test by one that tests that. And moved it to test_function because all this isn't about integrals; I'm not changing anything in the integrals module.

ylemkimon commented:
It should've been removed in #1162. +1

normalhuman committed:
Test the issue much more efficiently

normalhuman committed:
Move the test to test_function where it belongs

It was strange to test a change to functions.py in test_integrals.
The change is not specific to integrals at all.
